File: packages/sdk-ts/tests/runtimeCompatibility.test.ts
import { describe, expect, it } from "vitest";
import {
  negotiateRuntimeCompatibility,
  type RuntimeRequirement,
} from "../src/runtimeCompatibility.ts";

const requirement: RuntimeRequirement = {
  protocolVersion: "1.2.4",
};
const marker = (protocolVersion: string) => ({
  protocolVersion,
  serverInfo: { name: "stagehand", version: "1.0.0" },
});

describe("negotiateRuntimeCompatibility", () => {
  it("accepts an older patch on the same minor", () =>
    expect(negotiateRuntimeCompatibility(requirement, marker("1.2.0"))).toStrictEqual({
      kind: "compatible",
      protocolVersion: "1.2.0",
      serverInfo: { name: "stagehand", version: "1.0.0" },
    }));
  it("accepts a newer server minor", () =>
    expect(negotiateRuntimeCompatibility(requirement, marker("1.9.0"))).toMatchObject({
      kind: "compatible",
      protocolVersion: "1.9.0",
    }));
  it("reports a server minor below the client requirement", () =>
    expect(negotiateRuntimeCompatibility(requirement, marker("1.1.99"))).toMatchObject({
      kind: "incompatible",
      reason: "protocol-server-too-old",
      required: { protocolVersion: "1.2.4" },
      reported: {
        protocolVersion: "1.1.99",
        serverInfo: { name: "stagehand", version: "1.0.0" },
      },
    }));
  it("reports a protocol major mismatch", () =>
    expect(negotiateRuntimeCompatibility(requirement, marker("2.0.0"))).toMatchObject({
      kind: "incompatible",
      reason: "protocol-major-mismatch",
      required: { protocolVersion: "1.2.4" },
      reported: {
        protocolVersion: "2.0.0",
        serverInfo: { name: "stagehand", version: "1.0.0" },
      },
    }));
  it("accepts an exact prerelease and rejects a different prerelease", () => {
    const prereleaseRequirement = { protocolVersion: "1.3.0-beta.1" };
    expect(
      negotiateRuntimeCompatibility(prereleaseRequirement, marker("1.3.0-beta.1")),
    ).toMatchObject({ kind: "compatible" });
    expect(
      negotiateRuntimeCompatibility(prereleaseRequirement, marker("1.3.0-beta.2")),
    ).toMatchObject({ kind: "incompatible", reason: "protocol-prerelease-mismatch" });
  });
  it("reports an invalid client requirement without throwing", () =>
    expect(
      negotiateRuntimeCompatibility({ protocolVersion: "not-semver" }, marker("1.2.4")),
    ).toMatchObject({
      kind: "incompatible",
      reason: "protocol-invalid-version",
    }));
  it("reports a non-SemVer reported protocol version as incompatible, not unknown", () =>
    expect(negotiateRuntimeCompatibility(requirement, marker("not-semver"))).toMatchObject({
      kind: "incompatible",
      reason: "protocol-invalid-version",
      detail: "Invalid protocol version: client 1.2.4, server not-semver",
      reported: {
        protocolVersion: "not-semver",
        serverInfo: { name: "stagehand", version: "1.0.0" },
      },
    }));
  it.each([
    [{ ...marker("1.2.4"), serverInfo: { name: "", version: "1.0.0" } }],
    [{ ...marker("1.2.4"), serverInfo: { name: "stagehand", version: "" } }],
    [{ protocolVersion: "", serverInfo: { name: "stagehand", version: "1.0.0" } }],
  ])("reports a marker with an empty field as unreadable for %j", (raw) =>
    expect(negotiateRuntimeCompatibility(requirement, raw)).toMatchObject({
      kind: "unknown",
      reason: "unreadable-marker",
    }),
  );
  it.each([[null], [undefined]])("reports a missing marker for %s", (raw) =>
    expect(negotiateRuntimeCompatibility(requirement, raw)).toMatchObject({
      kind: "unknown",
      reason: "missing-marker",
      detail: "Runtime marker is absent",
    }),
  );
  it.each([[0], ["x"], [[]], [{ protocolVersion: "1" }], [{ protocolVersion: 1 }]])(
    "reports an unreadable malformed marker for %j",
    (raw) =>
      expect(negotiateRuntimeCompatibility(requirement, raw)).toMatchObject({
        kind: "unknown",
        reason: "unreadable-marker",
      }),
  );
  it("reports a foreign runtime as incompatible", () =>
    expect(
      negotiateRuntimeCompatibility(requirement, {
        ...marker("1.2.4"),
        serverInfo: { name: "other", version: "1.0.0" },
      }),
    ).toMatchObject({
      kind: "incompatible",
      reason: "runtime-name-mismatch",
      detail: 'Runtime name mismatch: expected "stagehand", server reported "other"',
      required: { protocolVersion: "1.2.4" },
      reported: {
        protocolVersion: "1.2.4",
        serverInfo: { name: "other", version: "1.0.0" },
      },
    }));
  it("reports unknown marker keys as unreadable", () =>
    expect(
      negotiateRuntimeCompatibility(requirement, { ...marker("1.2.4"), status: "ready" }),
    ).toMatchObject({
      kind: "unknown",
      reason: "unreadable-marker",
    }));
  it("does not throw for an unreadable proxy", () => {
    const raw = new Proxy({}, { get: () => throwOnRead() });
    expect(negotiateRuntimeCompatibility(requirement, raw)).toMatchObject({
      kind: "unknown",
      reason: "unreadable-marker",
    });
  });
  it("is deterministic and does not mutate inputs", () => {
    const required = { ...requirement };
    const reported = marker("1.2.4");
    const before = structuredClone({ required, reported });
    expect(negotiateRuntimeCompatibility(required, reported)).toEqual(
      negotiateRuntimeCompatibility(required, reported),
    );
    expect({ required, reported }).toEqual(before);
  });
});

function throwOnRead(): never {
  throw new Error("unreadable");
}

File: packages/sdk-go/runtime_compatibility_test.go
package stagehand

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
	"testing"
)

func TestNegotiateRuntimeCompatibility(t *testing.T) {
	t.Parallel()

	protocolMajor, err := strconv.Atoi(strings.Split(stagehandProtocolVersion, ".")[0])
	if err != nil {
		t.Fatal(err)
	}
	incompatibleProtocolVersion := fmt.Sprintf("%d.0.0", protocolMajor+1)

	tests := []struct {
		name       string
		marker     string
		compatible bool
		detail     string
	}{
		{
			name: "compatible",
			marker: fmt.Sprintf(`{
				"protocolVersion": %q,
				"serverInfo": {"name": "stagehand", "version": "1.0.0"}
			}`, stagehandProtocolVersion),
			compatible: true,
			detail:     "protocolVersion=" + stagehandProtocolVersion,
		},
		{
			name:       "missing marker",
			marker:     `null`,
			compatible: false,
			detail:     "no Stagehand runtime marker",
		},
		{
			name: "major mismatch",
			marker: fmt.Sprintf(`{
				"protocolVersion": %q,
				"serverInfo": {"name": "stagehand", "version": "1.0.0"}
			}`, incompatibleProtocolVersion),
			compatible: false,
			detail:     "major mismatch",
		},
		{
			name: "wrong runtime",
			marker: fmt.Sprintf(`{
				"protocolVersion": %q,
				"serverInfo": {"name": "other", "version": "1.0.0"}
			}`, stagehandProtocolVersion),
			compatible: false,
			detail:     `serverInfo.name="other"`,
		},
		{
			name: "invalid protocol version",
			marker: `{
				"protocolVersion": 1,
				"serverInfo": {"name": "stagehand", "version": "1.0.0"}
			}`,
			compatible: false,
			detail:     "protocolVersion=1",
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			t.Parallel()

			compatible, detail := negotiateRuntimeCompatibility(
				json.RawMessage(test.marker),
			)
			if compatible != test.compatible {
				t.Fatalf("compatible = %t, want %t", compatible, test.compatible)
			}
			if !strings.Contains(detail, test.detail) {
				t.Fatalf("detail = %q, want it to contain %q", detail, test.detail)
			}
		})
	}
}

func TestProtocolCompatibility(t *testing.T) {
	t.Parallel()

	for _, test := range []struct {
		name       string
		client     string
		server     string
		compatible bool
		detail     string
	}{
		{name: "patch ignored", client: "1.2.4", server: "1.2.0", compatible: true},
		{name: "newer server minor", client: "1.2.4", server: "1.9.0", compatible: true},
		{name: "server too old", client: "1.2.4", server: "1.1.99", detail: "older"},
		{name: "major mismatch", client: "1.2.4", server: "2.0.0", detail: "major mismatch"},
		{name: "exact prerelease", client: "1.3.0-beta.1", server: "1.3.0-beta.1", compatible: true},
		{name: "different prerelease", client: "1.3.0-beta.1", server: "1.3.0-beta.2", detail: "match exactly"},
		{name: "invalid client", client: "not-semver", server: "1.3.0", detail: "invalid protocol version"},
		{name: "invalid server", client: "1.3.0", server: "not-semver", detail: "invalid protocol version"},
	} {
		t.Run(test.name, func(t *testing.T) {
			t.Parallel()
			compatible, detail := protocolCompatibility(test.client, test.server)
			if compatible != test.compatible {
				t.Fatalf("compatible = %t, want %t", compatible, test.compatible)
			}
			if !strings.Contains(detail, test.detail) {
				t.Fatalf("detail = %q, want it to contain %q", detail, test.detail)
			}
		})
	}
}
